### PR TITLE
add lock for multi-threaded directory creation

### PR DESCRIPTION
**Pull request by ionox0**
@mr-c This has fixed the issue for me, let me know your thoughts.

However it's mostly just to show the issue, because I haven't added any tests for it. 

It seems that the lock needs to be passed into the thread explicitly, I'm no expert with locking in Python, but does that check out with your understanding of how it works?

https://github.com/common-workflow-language/cwltool/pull/1037
https://github.com/common-workflow-language/cwltool/issues/1036

### CODE REVIEW

File: cwltool/job.py [hunk 1]
```diff
@@ -414,11 +414,17 @@ def get_tree_mem_usage(memory_usage):
...
 class CommandLineJob(JobBase):
     def run(self,
-            runtimeContext  # type: RuntimeContext
+            runtimeContext,         # type: RuntimeContext
+            TMPDIR_LOCK=None        # type: threading.Lock
             ):  # type: (...) -> None
 
-        if not os.path.exists(self.tmpdir):
-            os.makedirs(self.tmpdir)
+        if TMPDIR_LOCK != None:
```
Comment (ionox0) [inline cwltool/job.py]: I did this becasue I'm not passing the lock in from the SingleThreadExecutor class

File: cwltool/job.py [hunk 2]
```diff
@@ -414,11 +414,17 @@ def get_tree_mem_usage(memory_usage):
 
 class CommandLineJob(JobBase):
     def run(self,
-            runtimeContext  # type: RuntimeContext
+            runtimeContext,         # type: RuntimeContext
+            TMPDIR_LOCK=None        # type: threading.Lock
```
Comment (mr-c) [inline cwltool/job.py]:
```suggestion
            tmpdirLock=None        # type: threading.Lock
```

File: cwltool/job.py (hunk 1 above)
Comment (mr-c) [reply cwltool/job.py]:
```suggestion
        if tmpdirLock != None:
```

File: cwltool/job.py [hunk 3]
```diff
@@ -414,11 +414,17 @@ def get_tree_mem_usage(memory_usage):
...
             ):  # type: (...) -> None
 
-        if not os.path.exists(self.tmpdir):
-            os.makedirs(self.tmpdir)
+        if TMPDIR_LOCK != None:
+            if not os.path.exists(self.tmpdir):
```
Comment (mr-c) [inline cwltool/job.py]: isn't this missing `with tmpdirLock:` before it?

**Comment by ionox0** [comment]
I addressed these three comments, however I'm seeing that there is a `JobBase` class that uses the `runtimeContext` which might be a better place for this `tmpdir_lock` to reside. 

I also see that the RuntimeContext has a `workflow_eval_lock`. I believe this would be a better option, rather than creating a new lock just for this tmpdir creation.

File: cwltool/job.py (hunk 3 above)
Comment (ionox0) [reply cwltool/job.py]: Yes, my mistake, I had included it in the version I was testing with but didn't commit it here.

**Comment by mr-c** [comment]
My understanding is that it is better for locks to have a narrow, specific purpose. Here we are trying to avoid contention around creating the same directory whereas `workflow_eval_lock` is for managing waiting for workflow components to finish.

**Comment by mr-c** [comment]
Jenkins, okay to test

File: cwltool/executors.py [hunk 4]
```diff
@@ -231,10 +234,10 @@ def select_resources(self, request, runtime_context):  # pylint: disable=unused-
 
         return result
 
-    def _runner(self, job, runtime_context):
+    def _runner(self, job, runtime_context, TMPDIR_LOCK):
         """ Job running thread. """
         try:
-            job.run(runtime_context)
+            job.run(runtime_context, TMPDIR_LOCK)
```
Comment (mr-c) [inline cwltool/executors.py]: `TypeError: run() takes 2 positional arguments but 3 were given`

**Comment by mr-c** [comment]
@ionox0 Can you enable edits to your branch from maintainers? https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by mr-c** [comment]
@ionox0 Nevermind, was my error

**Pull request merged**